source/dofs/dof_handler_policy.cc

```diff
@@ -5591,6 +5591,103 @@ namespace internal
         Assert(triangulation != nullptr, ExcInternalError());
 
 
+        // We start by checking whether only the numbering within the MPI
+        // ranks changed. In that case, we can apply the renumbering with some
+        // local renumbering only (this is similar to the renumber_mg_dofs()
+        // function below)
+        bool locally_owned_set_changes = false;
+        for (types::global_dof_index i : new_numbers)
+          if (dof_handler->locally_owned_dofs().is_element(i) == false)
+            {
+              locally_owned_set_changes = true;
```

Comment:
I think it would still be easier to convert `new_numbers` to an index set and then compare those. The conversion of an index set should (if `new_numbers` is sorted) be $O(n)$ and the comparison also at most, whereas repeatedly calling `is_element()` will be `O(N log(N))`.

Reply:
(As would sorting `new_numbers`, but I suspect that that would still be cheaper.)

Reply:
Let me try to convince you again: The current code is likely faster whenever the `break` one line below is hit. This is the case when we need to switch to the expensive path. On the cheap path, I agree that your suggestion after sorting (we must expect a renumbering to change something) will likely be faster because we do not need to go through the `lower_bound` checks in `is_element`. On the other hand, I expect the code to often be used with a contiguous range or at least not too many different ranges. So to summarize, we would have a more expensive operation in case the final result is `false`. But overall that case would be faster and much faster than the initialization. 

It's not that I don't want to change it, it's just that I do not see a big benefit and I want to impact the already more expensive path as little as possible.

Reply:
Good point, I didn't think of the quick exit case!